Remove jobs no longer referenced in the codebase

fredden opened the pull request:
### Description
When upgrading a Magento website over time, some cron jobs are left in the `cron_schedule` table for jobs which no longer exist. This can happen when a module is removed, or when a module removes a cronjob, or when a module changes the name of one of its cronjobs.

This pull request adds a daily clean-up task to remove such jobs from the schedule. This means that left-over jobs are pruned safely and the database table does have useless data.

### Manual testing scenarios
1. Run the Magento cron so that `sitemap_generate` gets scheduled. This belongs to the `Magento_Sitemap` module.
1. Disable the `Magento_Sitemap` module.
1. Run the Magento cron.
1. Observe that there is (or not) a left-over entry for `sitemap_generate` in the `cron_schedule` database table.

## Related issues

Fixes https://github.com/magento/magento2/issues/38217

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

fredden committed:
Remove jobs no longer referenced in the codebase

fredden commented:
@magento run all tests

kandy commented on app/code/Magento/Cron/Cron/CleanOldJobs.php:
```diff
@@ -0,0 +1,89 @@
+     *
+     * @return void
+     */
+    public function execute(): void
```
Can you clarify why we need the logic on top of https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php#L562

fredden replied:
@kandy the linked code only cleans up known jobs (see line 596) in known groups (see line 558). This pull request adds clean-up for unknown jobs / groups.

https://github.com/magento/magento2/blob/995a05c9c7bf661729d79a01ad1447370bf4f39f/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php#L596
https://github.com/magento/magento2/blob/995a05c9c7bf661729d79a01ad1447370bf4f39f/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php#L558

kandy replied:
ok, make sense 
thanks you for clarification

hostep commented:
Nice! Good job!

@fredden: can you add https://github.com/magento/magento2/issues/38217 to the description of this PR? As this fixes that issue, so it automatically gets closed when this PR is merged.

fredden commented:
>  add https://github.com/magento/magento2/issues/38217 to the description of this PR? As this fixes that issue

Yes, I'll update this later today. Thanks for highlighting; I hadn't seen that report before now.

> maybe we should add the cleanup of unknown jobs to the Recurring setup script that exists in the Cron module

I considered this approach too. This would catch most of the scenarios where groups or jobs are removed. There are two (main) reasons why I went with a periodic clean-up instead.

If the clean-up is in the `setup:...` process, then it should clean up all orphaned entries, not just "old" orphaned entries. This means that we immediately lose history for these jobs. When removing a module forever, this is typically fine; sufficient backing up of data is part of good maintenance. When a module simply renames a cronjob however (eg, to fix a typo), then losing history prematurely isn't ideal. (I'm thinking about the "timeline" view of a certain cronjobmanager module.)

The other reason for the periodic clean-up, is that I have seen modules dynamically create jobs, which then never get cleaned up by Magento when that module's configuration changes.

fredden commented:
@magento run all tests

fredden commented:
@magento run all tests

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented:
@magento run Database Compare, Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Unit Tests, WebAPI Tests

engcom-Hotel reviewed (changes requested):
Hello @fredden,

Thanks for the collaboration!

The changes looks good to me but it seems the Integration test has been failing due to changes. I request you to please check the same.

Thanks

fredden commented:
@engcom-Hotel thanks for your note. I don't know why I added a circular reference here. I've removed it now.

fredden commented:
@magento run all tests

engcom-Hotel commented:
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE, WebAPI Tests

fredden commented:
> I don't know why I added a circular reference here. I've removed it now.

I've worked out that this was a typing error. I had intended to add a dependency on `Magento_Config`. I've adjusted the commits here to reflect this.

fredden commented:
@magento run all tests

engcom-Hotel reviewed (approved):
Failed tests seems flaky to me.

engcom-Bravo commented:
Hi @fredden,

Thanks for the collaboration & contribution!

:heavy_check_mark: QA Passed

**Preconditions:**

- Install fresh Magento **`2.4-develop`**

**Steps to reproduce**

As per the Description.


**Before:** :heavy_multiplication_x: 


<img width="1436" alt="Screenshot 2024-05-24 at 12 29 38" src="https://github.com/magento/magento2/assets/51680745/61d16e48-726d-4965-be1c-5bafb17b38c7">





**After:** :heavy_check_mark:  


The Jobs for Disabled Modules are removed from the `cron_schedule` table from the database.

Builds are failed. Hence, moving this PR to **Extended Testing.**


Thanks.

engcom-Echo commented:
@magento run Functional Tests EE,Functional Tests CE,Functional Tests B2B

engcom-Echo commented:
Functional Tests B2B,Functional Tests CE,Functional Tests EE are different on last two run on same commit.
Also it contains some known failure.

[ACQE-6523](https://jira.corp.adobe.com/browse/ACQE-6523) - AdminReorderWithCatalogPriceRuleDiscountTest
[ACQE-6331](https://jira.corp.adobe.com/browse/ACQE-6331) - StorefrontCreateOrderAllQuantityGroupedProductOptionDefaultStockTest

Functional Tests B2B
Run1
<img width="1340" alt="38693-b2b" src="https://github.com/magento/magento2/assets/51681435/21c226d0-31ea-4416-9424-e04cd6cf6606">

Run2
<img width="915" alt="38693-b2b2" src="https://github.com/magento/magento2/assets/51681435/e394ddbc-4085-4a51-a7f4-4d36c18e3cac">

Functional Tests CE
Run1
<img width="1440" alt="38693-ce" src="https://github.com/magento/magento2/assets/51681435/d16e079e-a6e0-40cd-b4cd-80cffeab265e">

Run2
<img width="1141" alt="38693-ce2" src="https://github.com/magento/magento2/assets/51681435/e7ce4ebd-f893-4e5b-97b6-2df8bf03c9bb">

Functional Tests EE
Run1
<img width="1428" alt="38693-ee" src="https://github.com/magento/magento2/assets/51681435/ab6012e9-2778-4b63-9f89-644548f8a89a">

Run2
<img width="981" alt="38693-ee2" src="https://github.com/magento/magento2/assets/51681435/8f4fa549-df7c-40b4-8a58-178a35d82fa2">


Hence moving it to merge in progress

MaximGns commented:
Questioning why you should execute this by using another cron job? I guess cronjobs that are removed/changed name is only discoverd after a setup upgrade, and would be better do run this job after a setup upgrade (during deployment to production for example). Some project will not build anything to production for months, therefore no cronjobs will be removed/changed by name in the meantime...

fredden commented:
@MaximGns see https://github.com/magento/magento2/pull/38693#issuecomment-2092442675
